### PR TITLE
REST API: Meta update fails if unrelated unchanged field is a multi-item array

### PR DESCRIPTION
**Pull request by anton-vlasenko**
Trac ticket: https://core.trac.wordpress.org/ticket/60618

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/meta.php [hunk 1]
```diff
@@ -174,14 +174,19 @@ function add_metadata( $meta_type, $object_id, $meta_key, $meta_value, $unique =
...
+ * @param bool   $is_failure Optional. Set to true if updating metadata fails due to a database error.
+ *                           Set to false if the database query is successful.
+ *                           Set to null if no query is executed.
  * @return int|bool The new meta field ID if a field with the given key didn't exist
  *                  and was therefore added, true on successful update,
  *                  false on failure or if the value passed to the function
  *                  is the same as the one that is already in the database.
  */
-function update_metadata( $meta_type, $object_id, $meta_key, $meta_value, $prev_value = '' ) {
+function update_metadata( $meta_type, $object_id, $meta_key, $meta_value, $prev_value = '', &$is_failure = null ) {
```
Comment (spacedmonkey) [inline src/wp-includes/meta.php]: Would be easier, to get this function to return a WP_Error object, using a parameter. Like how `wp_insert_post` works?

Comment (anton-vlasenko) [reply src/wp-includes/meta.php]: Thank you for the review, @spacedmonkey.

I've been considering this option.  
`update_metadata()` is internally used in the following functions:
- `update_comment_meta()`  
- `update_site_meta()`
- `update_post_meta()`
- `update_term_meta()`
- `update_user_meta()`

Therefore, if the return value of `update_metadata()` changes, the return value of all these functions will also change.  
[According to this data](https://www.wpdirectory.net/search/01J9W80B80Z77X1EHRVXYSB3F9), there are 2 352 WordPress extensions that use return value of one or more of these functions.
As a result, all these plugins would need to be tested, and some might need to be refactored.  
Thus, I assume allowing `update_metadata()` to return a `WP_Error` object could break a lot of things.  

So, I decided to add an additional function argument, as this is the least disruptive way to detect when `update_metadata()`'s database request fails (as opposed to a request that simply didn’t update any database rows).

Comment (spacedmonkey) [reply src/wp-includes/meta.php]: @anton-vlasenko To be clear,  the parameter would be opt in, like this `bool $wp_error = false`. So no need to change the parmaters of  existing functions like this `update_comment_meta`. As the rest api meta api, use the raw functions, like `update_metadata` these functions can be opt-in. See https://core.trac.wordpress.org/ticket/15036

File: src/wp-includes/meta.php [hunk 2]
```diff
@@ -249,7 +251,13 @@ function update_metadata( $meta_type, $object_id, $meta_key, $meta_value, $prev_
 
 	$meta_ids = $wpdb->get_col( $wpdb->prepare( "SELECT $id_column FROM $table WHERE meta_key = %s AND $column = %d", $meta_key, $object_id ) );
 	if ( empty( $meta_ids ) ) {
-		return add_metadata( $meta_type, $object_id, $raw_meta_key, $passed_value );
+		$result = add_metadata( $meta_type, $object_id, $raw_meta_key, $passed_value );
+		// The return value of add_metadata() should also be checked, as it may indicate a failure.
+		if ( $wp_error && false === $result ) {
+			return new WP_Error( 'meta_add_error', __( 'Failed to add metadata.' ) );
```
Comment (anton-vlasenko) [inline src/wp-includes/meta.php:257]: I'm not sure what would be the best name for that error code.

File: src/wp-includes/meta.php [hunk 3]
```diff
@@ -306,6 +314,11 @@ function update_metadata( $meta_type, $object_id, $meta_key, $meta_value, $prev_
 	}
 
 	$result = $wpdb->update( $table, $data, $where );
+
+	if ( $wp_error && false === $result ) {
+		return new WP_Error( 'meta_update_error', __( 'Failed to update metadata.' ) );
```
Comment (anton-vlasenko) [inline src/wp-includes/meta.php:319]: Same as [above](https://github.com/WordPress/wordpress-develop/pull/7149/files#r1803533224).

File: src/wp-includes/meta.php (hunk 1 above)
Comment (anton-vlasenko) [reply src/wp-includes/meta.php]: Apologies, my mistake, @spacedmonkey. For some reason, I missed that the `wp_insert_post()` function already has a `$wp_error` parameter.

TL;DR: I agree with your suggestion and like the approach, especially since `wp_insert_post()` already has a similar mechanism in place.

Fixed in f556bfd857257ca3809b2505de02b95958476b1e.

Longer explanation: I had considered this approach previously but dismissed it, thinking that in the distant future, `update_metadata()` might need to return additional values (similar to how it should return a WP_Error object on database failure now).
This would potentially require two opt-in parameters, such as `$wp_error` and maybe `$wp_return_additional_data` or something similar.
With two opt-in parameters, it might become unclear which data to return if both are set to true. However, this is purely hypothetical and unlikely to occur.
